activation offloading implementation

xuanzhang816 opened the pull request:
This PR is for the **yellow block** of the design flow below.

<img width="3880" height="2240" alt="image" src="https://github.com/user-attachments/assets/1774b9c1-5a72-4ebe-9e5c-70072cc7f320" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream with correct waits for e.g., data dependencies. (Note that from a "frontend" perspective, everything is the same, the only difference is that the memcpy is now moved to a separate stream, which prepares for subsequent reordering and overlapping)

For an [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4), we have 
1. baseline (i.e., no activation offloading)

- 9160 MB peak memory
- 157.42 ms per-iteration runtime
- trace for fwd and bwd as below:
<img width="845" height="52" alt="image" src="https://github.com/user-attachments/assets/2868416e-e5a6-42a6-985f-891c5d8edb72" />

2. AO (i.e., with `enable_activation_offloading=True`)

- 7460 MB peak memory 
- 224.15 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="864" height="52" alt="image" src="https://github.com/user-attachments/assets/ec7b6326-b192-4915-af3f-824a2d3f7c49" />

3. AO on separate stream (i.e., additionally with `activation_offload_separate_stream = True`)

- 7460 MB peak memory 
- 224.11 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="891" height="121" alt="image" src="https://github.com/user-attachments/assets/2137691a-2c97-4208-80c6-cd7ae7fac7e0" />


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168316
* __->__ #167880



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

xuanzhang816 committed:
activation offloading implementation

[ghstack-poisoned]

Skylion007 commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,442 @@
+# mypy: allow-untyped-defs
```
Aren't all the defs typed? Just drop the content.

xuanzhang816 replied:
Thanks for catching this. Some AI wrote this and I missed it :P

xuanzhang816 commented:
@pytorchbot label "topic: not user facing"

xuanzhang816 committed:
Update on "activation offloading implementation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading implementation"

This PR is for the **yellow block** of the design flow below.

<img width="1940" height="1120" alt="image" src="https://github.com/user-attachments/assets/75e5972d-3b2f-4ac2-ad8f-da1493957243" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading implementation"

This PR is for the **yellow block** of the design flow below.

<img width="3880" height="2240" alt="image" src="https://github.com/user-attachments/assets/1774b9c1-5a72-4ebe-9e5c-70072cc7f320" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream with correct waits for e.g., data dependencies. (Note that from a "frontend" perspective, everything is the same, the only difference is that the memcpy is now moved to a separate stream, which prepares for subsequent reordering and overlapping)

For an [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4), we have 
1. baseline (i.e., no activation offloading)

- 9160 MB peak memory
- 157.42 ms per-iteration runtime
- trace for fwd and bwd as below:
<img width="845" height="52" alt="image" src="https://github.com/user-attachments/assets/2868416e-e5a6-42a6-985f-891c5d8edb72" />

2. AO (i.e., with `enable_activation_offloading=True`)

- 7460 MB peak memory 
- 224.15 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="864" height="52" alt="image" src="https://github.com/user-attachments/assets/ec7b6326-b192-4915-af3f-824a2d3f7c49" />

3. AO on separate stream (i.e., additionally with `activation_offload_separate_stream = True`)

- 7460 MB peak memory 
- 224.11 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="891" height="121" alt="image" src="https://github.com/user-attachments/assets/2137691a-2c97-4208-80c6-cd7ae7fac7e0" />





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela 

[ghstack-poisoned]

eellison commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,462 @@
+    Forward pass modification for CPU offloading, by inserting offloaindg nodes.
+
+    Args:
+        graph: The forward graph to modify
```
nit: not sure this Args: adds anything

eellison commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,462 @@
+            cpu_node: fx.Node = graph.call_function(
+                torch.ops.prims.device_put.default,
+                args=(node, torch.device("cpu")),
+                kwargs={"non_blocking": True},
```
just to verify: this will output in pinned memory, right

eellison commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,462 @@
+        if can_offload(node, fwd_outputs, model_outputs, static_lifetime_input_nodes):
+            node.meta["saved_for_offloading"] = True
+            node.meta["original_device"] = node.meta["val"].device
+            should_perform_offloading = True
```
so, we're offloading all activations today ?

xuanzhang816 replied:
we offload all activations that are explicitly marked, because I have these lines before this (i.e., skipping the loop if they are not marked explicitly). Let me add a comment to make this clear: 
```
        if node.meta.get("should_offload", False) is False:
            continue
```

> i imagine we might want to do something where the earlier activations in the backward are not offloaded

Agree with this, I think this should be more in the auto logic? Again, let me add a comment somewhere to make sure this is considered.

eellison commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,462 @@
+    offload_stream_id: int = new_stream()
+
+    for offload_node in offload_nodes:
+        offload_ready_event_id: int = new_event()
```
will have to defer to @mlazos on this one, same with `new_stream`.

eellison reviewed (approved):
Looks good ! My main questions are around :

1. if the stream forking behavior should be querying for current stream at runtime 
2. future policy for deciding when to offload.

cc @mlazos for review as well

xuanzhang816 commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,462 @@
+    effective_users_cache: dict[fx.Node, OrderedSet[fx.Node]] = {}
+    op_types: OpTypes = get_default_op_list()
+
+    def find_all_effective_users(node: fx.Node) -> OrderedSet[fx.Node]:
```
Very good question, I think you are right, we can offload immediately and have two copies, and the one in GPU memory will be removed once all the effective users are done computing.

xuanzhang816 commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,462 @@
+    op_types: OpTypes = get_default_op_list()
+
+    if node not in fwd_outputs:
+        log.debug("\tSkipped! Can only offload nodes in fwd_module_outputs.")
```
There is an line of debug logging before the checks, which is
```
    log.debug(f"Checking node {node.name} for offloading...") 
```
So we do have the node.name in the logs.

mlazos commented:
> Looks good ! My main questions are around :
> 
> 1. if the stream forking behavior should be querying for current stream at runtime
> 2. future policy for deciding when to offload.
> 
> cc @mlazos for review as well

It actually does - when you retrieve the current stream it gets an index which dynamo populates with a call to the current stream.

xuanzhang816 committed:
Update on "activation offloading implementation"

This PR is for the **yellow block** of the design flow below.

<img width="3880" height="2240" alt="image" src="https://github.com/user-attachments/assets/1774b9c1-5a72-4ebe-9e5c-70072cc7f320" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream with correct waits for e.g., data dependencies. (Note that from a "frontend" perspective, everything is the same, the only difference is that the memcpy is now moved to a separate stream, which prepares for subsequent reordering and overlapping)

For an [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4), we have 
1. baseline (i.e., no activation offloading)

- 9160 MB peak memory
- 157.42 ms per-iteration runtime
- trace for fwd and bwd as below:
<img width="845" height="52" alt="image" src="https://github.com/user-attachments/assets/2868416e-e5a6-42a6-985f-891c5d8edb72" />

2. AO (i.e., with `enable_activation_offloading=True`)

- 7460 MB peak memory 
- 224.15 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="864" height="52" alt="image" src="https://github.com/user-attachments/assets/ec7b6326-b192-4915-af3f-824a2d3f7c49" />

3. AO on separate stream (i.e., additionally with `activation_offload_separate_stream = True`)

- 7460 MB peak memory 
- 224.11 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="891" height="121" alt="image" src="https://github.com/user-attachments/assets/2137691a-2c97-4208-80c6-cd7ae7fac7e0" />





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela 

[ghstack-poisoned]

soulitzer commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -0,0 +1,497 @@
+        return
+
+    # Get default stream id and offload stream id
+    default_stream_id: int = get_current_stream(
```
nit: Maybe rename mentions of "default" stream to ambient/current stream (here and other places)?

xuanzhang816 committed:
Update on "activation offloading implementation"

This PR is for the **yellow block** of the design flow below.

<img width="3880" height="2240" alt="image" src="https://github.com/user-attachments/assets/1774b9c1-5a72-4ebe-9e5c-70072cc7f320" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream with correct waits for e.g., data dependencies. (Note that from a "frontend" perspective, everything is the same, the only difference is that the memcpy is now moved to a separate stream, which prepares for subsequent reordering and overlapping)

For an [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4), we have 
1. baseline (i.e., no activation offloading)

- 9160 MB peak memory
- 157.42 ms per-iteration runtime
- trace for fwd and bwd as below:
<img width="845" height="52" alt="image" src="https://github.com/user-attachments/assets/2868416e-e5a6-42a6-985f-891c5d8edb72" />

2. AO (i.e., with `enable_activation_offloading=True`)

- 7460 MB peak memory 
- 224.15 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="864" height="52" alt="image" src="https://github.com/user-attachments/assets/ec7b6326-b192-4915-af3f-824a2d3f7c49" />

3. AO on separate stream (i.e., additionally with `activation_offload_separate_stream = True`)

- 7460 MB peak memory 
- 224.11 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="891" height="121" alt="image" src="https://github.com/user-attachments/assets/2137691a-2c97-4208-80c6-cd7ae7fac7e0" />





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela 

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading implementation"

This PR is for the **yellow block** of the design flow below.

<img width="3880" height="2240" alt="image" src="https://github.com/user-attachments/assets/1774b9c1-5a72-4ebe-9e5c-70072cc7f320" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream with correct waits for e.g., data dependencies. (Note that from a "frontend" perspective, everything is the same, the only difference is that the memcpy is now moved to a separate stream, which prepares for subsequent reordering and overlapping)

For an [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4), we have 
1. baseline (i.e., no activation offloading)

- 9160 MB peak memory
- 157.42 ms per-iteration runtime
- trace for fwd and bwd as below:
<img width="845" height="52" alt="image" src="https://github.com/user-attachments/assets/2868416e-e5a6-42a6-985f-891c5d8edb72" />

2. AO (i.e., with `enable_activation_offloading=True`)

- 7460 MB peak memory 
- 224.15 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="864" height="52" alt="image" src="https://github.com/user-attachments/assets/ec7b6326-b192-4915-af3f-824a2d3f7c49" />

3. AO on separate stream (i.e., additionally with `activation_offload_separate_stream = True`)

- 7460 MB peak memory 
- 224.11 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="891" height="121" alt="image" src="https://github.com/user-attachments/assets/2137691a-2c97-4208-80c6-cd7ae7fac7e0" />





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela 

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading implementation"

This PR is for the **yellow block** of the design flow below.

<img width="3880" height="2240" alt="image" src="https://github.com/user-attachments/assets/1774b9c1-5a72-4ebe-9e5c-70072cc7f320" />

Assuming users mark some nodes with metadata `should_offload` (which could later be replaced as compiler automatic decisions), the implementation first checks if the tensors are safe to offload via the `can_offload()` function. For tensors that are marked as `should_offload` and pass the verification in `can_offload()`, we modify the forward and backward graph, where in the forward graph, we offload tensors to CPU, and in the backward graph, we reload tensors to GPU. 

We introduce two flags:
* `enable_activation_offloading` -- with this turned on, we insert offload/reload nodes to the fwd/bwd graphs.
* `activation_offload_separate_stream` -- with this turned on, we wrap the offload/reload nodes in a separate stream with correct waits for e.g., data dependencies. (Note that from a "frontend" perspective, everything is the same, the only difference is that the memcpy is now moved to a separate stream, which prepares for subsequent reordering and overlapping)

For an [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4), we have 
1. baseline (i.e., no activation offloading)

- 9160 MB peak memory
- 157.42 ms per-iteration runtime
- trace for fwd and bwd as below:
<img width="845" height="52" alt="image" src="https://github.com/user-attachments/assets/2868416e-e5a6-42a6-985f-891c5d8edb72" />

2. AO (i.e., with `enable_activation_offloading=True`)

- 7460 MB peak memory 
- 224.15 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="864" height="52" alt="image" src="https://github.com/user-attachments/assets/ec7b6326-b192-4915-af3f-824a2d3f7c49" />

3. AO on separate stream (i.e., additionally with `activation_offload_separate_stream = True`)

- 7460 MB peak memory 
- 224.11 ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="891" height="121" alt="image" src="https://github.com/user-attachments/assets/2137691a-2c97-4208-80c6-cd7ae7fac7e0" />





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78 

[ghstack-poisoned]
